packages/cw721/src/msg.rs

```diff
@@ -201,9 +288,513 @@ pub enum Cw721MigrateMsg {
+        }
+    }
+}
+
+/// This is a wrapper around CollectionInfo that includes the extension.
+#[cw_serde]
+pub struct CollectionInfoAndExtensionResponse<TCollectionExtension> {
     pub name: String,
     pub symbol: String,
+    pub extension: TCollectionExtension,
+    pub updated_at: Timestamp,
+}
+
+/// This is a wrapper around CollectionInfo that includes the extension, contract info, and number of tokens (supply).
+#[cw_serde]
+pub struct AllInfoResponse {
+    // contract details
+    pub contract_info: ContractInfoResponse,
+    // collection details
+    pub collection_info: CollectionInfo,
+    pub collection_extension: CollectionExtensionAttributes,
+    // NFT details
+    pub num_tokens: u64,
+}
+
+impl<T> From<CollectionInfoAndExtensionResponse<T>> for CollectionInfo {
+    fn from(response: CollectionInfoAndExtensionResponse<T>) -> Self {
+        CollectionInfo {
+            name: response.name,
+            symbol: response.symbol,
+            updated_at: response.updated_at,
+        }
+    }
+}
+
+impl<TCollectionExtension, TCollectionExtensionMsg>
+    StateFactory<CollectionInfoAndExtensionResponse<TCollectionExtension>>
+    for CollectionInfoMsg<TCollectionExtensionMsg>
+where
+    TCollectionExtension: Cw721State,
+    TCollectionExtensionMsg: Cw721CustomMsg + StateFactory<TCollectionExtension>,
+{
+    fn create(
+        &self,
+        deps: Option<Deps>,
+        env: Option<&Env>,
+        info: Option<&MessageInfo>,
+        current: Option<&CollectionInfoAndExtensionResponse<TCollectionExtension>>,
+    ) -> Result<CollectionInfoAndExtensionResponse<TCollectionExtension>, Cw721ContractError> {
+        self.validate(deps, env, info, current)?;
+        match current {
+            // Some: update existing metadata
+            Some(current) => {
+                let mut updated = current.clone();
+                if let Some(name) = &self.name {
+                    updated.name = name.clone();
+                }
+                if let Some(symbol) = &self.symbol {
+                    updated.symbol = symbol.clone();
+                }
+                let current_extension = current.extension.clone();
+                let updated_extension =
+                    self.extension
+                        .create(deps, env, info, Some(&current_extension))?;
+                updated.extension = updated_extension;
+                Ok(updated)
+            }
+            // None: create new metadata
+            None => {
+                let extension = self.extension.create(deps, env, info, None)?;
+                let env = env.ok_or(Cw721ContractError::NoEnv)?;
+                let new = CollectionInfoAndExtensionResponse {
+                    name: self.name.clone().unwrap(),
+                    symbol: self.symbol.clone().unwrap(),
+                    extension,
+                    updated_at: env.block.time,
+                };
+                Ok(new)
+            }
+        }
+    }
```

Comment:
And here's the magic, main logic being used for both, instantiation and execution: `StateFactory.create()`. In first line it always call `validate()` and checks whether sender is creator.

As noted earlier by shane, we may rename `StateFactory` -> `StateController`.